packages/block-editor/src/components/inserter/block-types-tab.js

```diff
@@ -101,82 +85,146 @@ export function BlockTypesTab(
+							items={ categoryItems }
 							onSelect={ onSelectItem }
 							onHover={ onHover }
-							label={ _x( 'Most used', 'blocks' ) }
+							label={ category.title }
 						/>
 					</InserterPanel>
-				) }
-
-				{ currentlyRenderedCategories.map( ( category ) => {
-					const categoryItems = itemsPerCategory[ category.slug ];
-					if ( ! categoryItems || ! categoryItems.length ) {
+				);
+			} ) }
+
+			{ didRenderAllCategories && uncategorizedItems.length > 0 && (
+				<InserterPanel
+					className="block-editor-inserter__uncategorized-blocks-panel"
+					title={ __( 'Uncategorized' ) }
+				>
+					<BlockTypesList
+						items={ uncategorizedItems }
+						onSelect={ onSelectItem }
+						onHover={ onHover }
+						label={ __( 'Uncategorized' ) }
+					/>
+				</InserterPanel>
+			) }
+
+			{ currentlyRenderedCollections.map(
+				( [ namespace, collection ] ) => {
+					const collectionItems = itemsPerCollection[ namespace ];
+					if ( ! collectionItems || ! collectionItems.length ) {
 						return null;
 					}
+
 					return (
 						<InserterPanel
-							key={ category.slug }
-							title={ category.title }
-							icon={ category.icon }
+							key={ namespace }
+							title={ collection.title }
+							icon={ collection.icon }
 						>
 							<BlockTypesList
-								items={ categoryItems }
+								items={ collectionItems }
 								onSelect={ onSelectItem }
 								onHover={ onHover }
-								label={ category.title }
+								label={ collection.title }
 							/>
 						</InserterPanel>
 					);
-				} ) }
+				}
+			) }
+		</>
+	);
+}
 
-				{ didRenderAllCategories && uncategorizedItems.length > 0 && (
-					<InserterPanel
-						className="block-editor-inserter__uncategorized-blocks-panel"
-						title={ __( 'Uncategorized' ) }
-					>
-						<BlockTypesList
-							items={ uncategorizedItems }
-							onSelect={ onSelectItem }
+export function BlockTypesTab(
+	{ rootClientId, onInsert, onHover, showMostUsedBlocks },
+	ref
+) {
+	const [ items, categories, collections, onSelectItem ] = useBlockTypesState(
+		rootClientId,
+		onInsert
+	);
+	const rootBlockName = useSelect(
+		( select ) => {
+			return select( blockEditorStore ).getBlockName( rootClientId );
+		},
+		[ rootClientId ]
+	);
+
+	if ( ! items.length ) {
+		return <InserterNoResults />;
+	}
+
+	const itemsForCurrentRoot = [];
+	const itemsRemaining = [];
+
+	for ( const item of items ) {
+		if ( rootClientId && item.rootClientId === rootClientId ) {
+			itemsForCurrentRoot.push( item );
+		} else {
+			itemsRemaining.push( item );
+		}
+	}
+
+	return (
+		<InserterListbox>
+			<div ref={ ref }>
+				{ !! itemsForCurrentRoot.length && (
+					<>
+						<div
+							className="block-editor-inserter__panel-header"
+							style={ { display: 'block' } }
+						>
+							<h2 className="block-editor-inserter__panel-title">
+								{ __( 'Can be inserted into ' ) +
+									getBlockType( rootBlockName )?.title }
+							</h2>
+						</div>
```

Comment:
Btw, this needs eyes from Design, I don't like the text and formatting here, it should be polished. Additionally, we need a heading in the "remaining block" section as well to make it clearer to the user.

Reply:
This is exemplified in the issue [here](https://github.com/WordPress/gutenberg/issues/60991#issuecomment-2114746204)

Reply:
It's too simplistic imo. You can have a block selected where you can have many block across multiple categories than can be inserted, such as navigation. Imo it's better to have two sections, one with blocks that can be inserted right at the selection, and one for the remaining blocks that we'll insert in a parent.